Mathlib/RingTheory/Binomial.lean

```diff
@@ -9,60 +9,63 @@ import Mathlib.RingTheory.Polynomial.Pochhammer
 /-!
 # Binomial semirings and binomial rings
 
-In this file we introduce binomial semirings and binomial rings, together with related functions
-like binomial coefficients.
+In this file we introduce the binomial property as a Prop-valued mixin, and define the `multichoose`
+and `choose` functions generalizing binomial coefficients.
 
 According to our main reference [elliott2006binomial] (which lists many equivalent conditions), a
 binomial ring is a torsion-free commutative ring `R` such that for any `x ∈ R` and any `k ∈ ℕ`, the
-product `x(x-1)⋯(x-k+1)` is divisible by `k!`.  The torsion-free condition lets us divide by `k!`
+product `x(x-1)⋯(x-k+1)` is divisible by `k!`. The torsion-free condition lets us divide by `k!`
 unambiguously, so we get uniquely defined binomial coefficients.
 
-The defining condition doesn't require commutativity, and we get a theory with essentially the same
-power by replacing subtraction with addition.  Thus, we consider a semiring `R` in which
-multiplication by factorials is injective, and demand that for all `x ∈ R` and any `k ∈ ℕ`,
-`x(x+1)⋯(x+(k-1))` is divisible by `k!`.
-The quotient is called `multichoose r k`, following the convention given for natural numbers.
+The defining condition doesn't require commutativity (or associativity), and we get a theory with
+essentially the same power by replacing subtraction with addition. Thus, we consider a semiring `R`
+in which multiplication by factorials is injective, and demand that the evaluation of the ascending
+Pochhammer polynomial `X(X+1)⋯(X+(k-1))` at any element is divisible by `k!`. The quotient is called
+`multichoose r k`, following the convention given for natural numbers.
 
 ## References
+
 * [J. Elliott, *Binomial rings, integer-valued polynomials, and λ-rings*][elliott2006binomial]
 
 -/
 
-universe u
+open Function
 
-variable {R : Type u}
+variable {R : Type*}
 
-section BinomialSemiring
+section Binomial
 
-/-- A semiring is binomial if multiplication by nonzero natural numbers is injective and ascending
-factorials are divisible by the corresponding factorial. -/
-class BinomialSemiring (R : Type u) extends Semiring R where
+/-- A `Prop`-valued mixin for multi-binomial coefficients. -/
+class Binomial (R : Type*) [Semiring R] : Prop where
   /-- Multiplication by positive integers is injective -/
-  nsmul_injective (n : ℕ) (r s : R) (h : n ≠ 0) : n • r = n • s → r = s
-  /-- The multichoose function witnesses the divisibility of ascFactorial r n by n! -/
-  multichoose : R → ℕ → R
-  /-- ascFactorial r n is divisible by n! (witnessed by multichoose) -/
-  factorial_nsmul_multichoose : ∀ (r : R) (n : ℕ),
-    n.factorial • multichoose r n = Polynomial.eval r (ascPochhammer R n)
+  nsmul_right_injective (n : ℕ) (h : n ≠ 0) : Injective (n • · : R → R)
+  /-- `ascFactorial r n` is divisible by n! -/
+  factorial_nsmul_multichoose (r : R) (n : ℕ) : ∃ (s : R),
+    n.factorial • s = Polynomial.eval r (ascPochhammer R n)
```

Comment:
I liked your previous idea better, actually. It let us actually compute `multichoose`.